lib/util/DateTimeHelpers.ts

```diff
@@ -0,0 +1,191 @@
+  return num.toLocaleString(undefined, { minimumIntegerDigits: min, useGrouping: false });
+}
+
+// Parsers:
+export function dateTimeParser(dateTimeStr: string, errorCreator?: () => Error): IDateTimeRepresentation {
+  // https://www.w3.org/TR/xmlschema-2/#dateTime
+  const [ date, time ] = dateTimeStr.split('T');
+  return { ...dateParser(date, errorCreator), ...internalTimeParser(time, errorCreator) };
+}
+
+export function dateTimeSerializer(date: IDateTimeRepresentation): string {
+  // Need to extract the date because we don't want the timeZone part here
+  return `${dateSerializer({ year: date.year, month: date.month, day: date.day })}T${timeSerializer(date)}`;
+}
+
+function timeZoneParser(timeZoneStr: string, errorCreator?: () => Error): Partial<ITimeZoneRepresentation> {
+  // https://www.w3.org/TR/xmlschema-2/#dateTime-timezones
+  if (timeZoneStr === '') {
+    return { zoneHours: undefined, zoneMinutes: undefined };
+  }
+  if (timeZoneStr === 'Z') {
+    return { zoneHours: 0, zoneMinutes: 0 };
+  }
+  const timeZoneStrings = timeZoneStr.replace(/^([+|-])(\d\d):(\d\d)$/gu, '$11!$2!$3').split('!');
+  const timeZone = timeZoneStrings.map(str => Number(str));
+  return {
+    zoneHours: timeZone[0] * timeZone[1],
+    zoneMinutes: timeZone[0] * timeZone[2],
+  };
+}
+
+function timeZoneSerializer(tz: Partial<ITimeZoneRepresentation>): string {
+  if (tz.zoneHours === undefined || tz.zoneMinutes === undefined) {
+    return '';
+  }
+  if (tz.zoneHours === 0 && tz.zoneMinutes === 0) {
+    return 'Z';
+  }
+  return `${tz.zoneHours >= 0 ? `+${numSer(tz.zoneHours)}` : numSer(tz.zoneHours)}:${numSer(Math.abs(tz.zoneMinutes))}`;
+}
+
+export function dateParser(dateStr: string, errorCreator?: () => Error): IDateRepresentation {
+  // https://www.w3.org/TR/xmlschema-2/#date-lexical-representation
+  const formatted = dateStr.replace(
+    /^(-)?([123456789]*\d{4})-(\d\d)-(\d\d)(Z|([+-]\d\d:\d\d))?$/gu, '$11!$2!$3!$4!$5',
+  );
+  if (formatted === dateStr) {
+    throw new Error('nono');
+  }
+  const dateStrings = formatted.split('!');
+  const date = dateStrings.slice(0, -1).map(str => Number(str));
+
+  const res = {
+    year: date[0] * date[1],
+    month: date[2],
+    day: date[3],
+    ...timeZoneParser(dateStrings[4]),
+  };
+  if (!(1 <= res.month && res.month <= 12) || !(1 <= res.day && res.day <= maximumDayInMonthFor(res.year, res.month))) {
+    throw new Error('nono');
+  }
+  return res;
+}
+
+export function rawTimeZoneExtractor(zoneContained: string): string {
+  const extraction = /(Z|([+-]\d\d:\d\d))?$/u.exec(zoneContained);
+  // It is safe to cast here because the empty string can always match.
+  return extraction![0];
+}
+
+export function dateSerializer(date: IDateRepresentation): string {
+  return `${numSer(date.year, 4)}-${numSer(date.month)}-${numSer(date.day)}${timeZoneSerializer(date)}`;
+}
+
+function internalTimeParser(timeStr: string, errorCreator?: () => Error): ITimeRepresentation {
```

Comment:
Is it really "internal" or is it just a time-specific parser?

Reply:
It's a time parser that parses time with accordance to the spec. XSD:Time however still requires a `%24` operation on the hours so in that sense it is not a time parser.

Reply:
> It's a time parser that parses time with accordance to the spec

Which spec is that? Can we refer to that spec, instead of using `internal`?

Reply:
This like is referenced in that function: https://www.w3.org/TR/xmlschema-2/#time-lexical-repr
Maybe the name should be `sharedParseTime` since it preforms the parse for multiple both `dateTime` and `time` but the `time` just needs one additional step (`% 24` on the hour field).

Reply:
I think I understand the confusion. I was also wondering why it wasn't called just `timeParser`, but I assume you did it because it's a helper function reused by both `time` and `dateTime` functions. What you really want is to not export this function out of the file, but that's hard, so you just name it something to avoid polluting the namespace.
I think convention for this would be `__parseTime` or `privateParseTime`.

Reply:
Yes, exactly! I changed it to `__parseTime`.